{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get list of papers from venue that cites a dataset"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import libs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To request OpenAlex\n",
    "import requests\n",
    "import numpy as np\n",
    "\n",
    "#Better print of dict\n",
    "from pprint import pprint\n",
    "\n",
    "#To load data\n",
    "import csv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load list of datasets and list of venues\n",
    "We will search for paper from venue contained in [venues.csv](../../data/venues.csv) referencing one or more of the datasets contained in [datasets.csv](../../data/datasets.csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Dictionnary with dataset's name as key and DOI as value\n",
    "datasets_doi = {}\n",
    "ds_reader = csv.DictReader(open('../../data/datasets.csv'))\n",
    "for ds in ds_reader:\n",
    "    datasets_doi[ds[\"name\"]] = ds[\"DOI\"]\n",
    "\n",
    "#Dictionnary with venues name as key and openalex id as value loaded from venues.csv\n",
    "venue_id = {}\n",
    "ds_reader = csv.DictReader(open('../../data/venues.csv'))\n",
    "for ds in ds_reader:\n",
    "    venue_id[ds[\"name\"]] = ds[\"openalex_id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Datasets:\n",
      "{'ACDC': '10.1109/TMI.2018.2837502',\n",
      " 'BRATS': '10.1109/tmi.2014.2377694',\n",
      " 'I2CVB': '10.1016/j.compbiomed.2015.02.009',\n",
      " 'LA': '10.1016/j.media.2020.101832',\n",
      " 'M&Ms': '10.1109/tmi.2021.3090082',\n",
      " 'MSCMRSeg': '10.48550/arxiv.2006.12434',\n",
      " 'Medical Decathlon': '10.1038/s41467-022-30695-9',\n",
      " 'PROMISE12': '10.1016/j.media.2013.12.002',\n",
      " 'Synapse': '10.7303/syn3193805'}\n",
      "\n",
      "Venues:\n",
      "{'LNCS': 'S106296714'}\n"
     ]
    }
   ],
   "source": [
    "print(\"Datasets:\")\n",
    "pprint(datasets_doi)\n",
    "print(\"\\nVenues:\")\n",
    "pprint(venue_id)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Try to convert DOI to OpenAlex ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Couldn't convert DOI for Synapse into OpenAlex ID\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Convert a DOI to OpenAlex ID used as value in some API field such as \"referenced_works\"\n",
    "@param\n",
    "    - DOI: the doi we want to convert\n",
    "@return\n",
    "    The OpenAlex ID if the DOI is in OpenAlex database, None otherwise\n",
    "\"\"\"\n",
    "def doi_to_OpenAlexId(doi):\n",
    "    base_url = f\"https://api.openalex.org/works/doi:{doi}\"\n",
    "    r = requests.get(base_url)\n",
    "    if r.status_code == 200:\n",
    "        r_json = r.json()\n",
    "        return r_json[\"id\"]\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "#Dictionnary with dataset names as key and openalex id as value. We associate an openalex ID because it's the value in the \"referenced_works\" field given by the API.\n",
    "datasets_id = {}\n",
    "\n",
    "#Convert DOI to OpenAlexID\n",
    "for ds  in datasets_doi:\n",
    "    openalex_id = doi_to_OpenAlexId(datasets_doi[ds])\n",
    "    if not openalex_id:\n",
    "        print(f\"Couldn't convert DOI for {ds} into OpenAlex ID\")\n",
    "    datasets_id[ds]=openalex_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ACDC': 'https://openalex.org/W2804047627',\n",
      " 'BRATS': 'https://openalex.org/W1641498739',\n",
      " 'I2CVB': 'https://openalex.org/W2049522781',\n",
      " 'LA': 'https://openalex.org/W3093394156',\n",
      " 'M&Ms': 'https://openalex.org/W4226199676',\n",
      " 'MSCMRSeg': 'https://openalex.org/W4312016581',\n",
      " 'Medical Decathlon': 'https://openalex.org/W3172681723',\n",
      " 'PROMISE12': 'https://openalex.org/W2106033751',\n",
      " 'Synapse': None}\n"
     ]
    }
   ],
   "source": [
    "pprint(datasets_id)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "note: Synapse value is None meaning it's note in OpenAlex database, therefore we won't be able to find references to it."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get list of papers from each venues citing at least one of the dataset"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Query OpenAlex filtering on the dataset and the venue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Dictionnary containing with dataset as key and a list of papers referencing the dataset as value\n",
    "paper_referencing = {ds:[] for ds in datasets_id}\n",
    "\n",
    "\n",
    "for ds in datasets_id:\n",
    "    for venue in venue_id:\n",
    "        #The list of paper referencing the dataset is decomposed in multiple pages so we have to iterate with the query parameter \"page\" to get them all.\n",
    "        next_page = True\n",
    "        page_number = 1\n",
    "        while next_page:\n",
    "            #Definition of the request\n",
    "            base_url = \"https://api.openalex.org/works\"\n",
    "            query_param = {\n",
    "                \"filter\":f\"cites:{datasets_id[ds]},locations.source.id:{venue_id[venue]}\",\n",
    "                \"page\":page_number\n",
    "            }\n",
    "            request = requests.get(base_url,params=query_param)\n",
    "\n",
    "            if request.status_code == 200:\n",
    "                request_json = request.json()\n",
    "                \n",
    "                #For each paper referencing the dataset we get the title (with a little transformation to remove \",\" and \"\\n\" inside of them), doi and publication year\n",
    "                for res in request_json[\"results\"]:\n",
    "                    title = res[\"title\"]\n",
    "                    title = title.replace(\",\",\"\")\n",
    "                    title = title.replace(\"\\n\",\"\")\n",
    "\n",
    "                    #Remove review paper\n",
    "                    if \"review\" in title.lower():\n",
    "                        continue\n",
    "\n",
    "                    doi = res[\"doi\"]\n",
    "                    if doi is not None:\n",
    "                        doi = doi[16:] #Remove the https://doi.org/\n",
    "                        paper_referencing[ds].append((title,doi,res[\"publication_year\"],res[\"abstract_inverted_index\"]))\n",
    "\n",
    "                #If the results field is empty that mean we are at the last page so we can continue to the next dataset\n",
    "                #otherwise we need to go to next page of the current dataset\n",
    "                if not request_json[\"results\"]:\n",
    "                    next_page = False\n",
    "                else:\n",
    "                    page_number += 1\n",
    "            else:\n",
    "                next_page = False\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of citations for ACDC: 120\n",
      "Number of citations for LA: 10\n",
      "Number of citations for MSCMRSeg: 0\n",
      "Number of citations for M&Ms: 34\n",
      "Number of citations for PROMISE12: 29\n",
      "Number of citations for Medical Decathlon: 4\n",
      "Number of citations for I2CVB: 13\n",
      "Number of citations for BRATS: 494\n",
      "Number of citations for Synapse: 0\n"
     ]
    }
   ],
   "source": [
    "for d in paper_referencing:\n",
    "    print(f\"Number of citations for {d}: {len(paper_referencing[d])}\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try to filter out wrong references\n",
    "This part of the notebook is to obtained a more accurate list of paper to only keep paper that are actually using the dataset and not only referencing the paper for another reason.\n",
    "\n",
    "To do that we will:\n",
    "1. Search for the dataset name in the abstract, making the hypothesis that if authors put the name of a dataset inside an abstract they must be using it.\n",
    "2. Look for the name of the dataset in Table of the fulltext, again if the dataset name is in a Table it should be used (especially as we removed review papers)\n",
    "3. Classify the figure in the fulltext to detect the organ of the cited dataset.\n",
    "\n",
    "If none of the above operation validate the citations, the paper is removed from the list. Otherwise, we can keep it.  "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  1. Check abstract for dataset's name"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will now query OpenAlex again for everypaper and search for the dataset name inside the \"abstract_inverted_index\" field of the API if it's present."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Reconstruct and transform the abstract of a paper using abstract_inverted_index field. We will removed non alpha numeric caracters and lower every word.\n",
    "\"\"\"\n",
    "def reconstruct_abstract(paper):\n",
    "    # Maximum size of the abstract, if the paper abstarct is longer it will be truncated\n",
    "    abstract = np.full(2500,\"\",dtype=object)\n",
    "    # The \"abstract_inverted_index\" field is a dictionnary with word as key and locations of this word in the abstract\n",
    "    # So we fill the abstarct variable above at the index of the word to reconstruct the abstract\n",
    "    if paper[3]:\n",
    "        for w in paper[3]:\n",
    "            for indices in paper[3][w]:\n",
    "                if indices < 2500:\n",
    "                    abstract[indices] = ''.join(filter(str.isalnum, w)).lower()\n",
    "        # Remove empty location mostly due to a shorter abstract \n",
    "        abstract = abstract[abstract != \"\"]\n",
    "        #Convert array to string\n",
    "    str_abstract = ' '.join(abstract)\n",
    "    return str_abstract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO\n",
    "paper_using_abstract = {dataset:[] for dataset in paper_referencing}\n",
    "for dataset in paper_referencing:\n",
    "    for paper in paper_referencing[dataset]:\n",
    "        abstract = reconstruct_abstract(paper)\n",
    "        if dataset.lower() in abstract:\n",
    "            paper_using_abstract[dataset].append(paper[:-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACDC': [('Learning to Segment Medical Images with Scribble-Supervision Alone',\n",
       "   '10.1007/978-3-030-00889-5_27',\n",
       "   2018),\n",
       "  ('Tripled-Uncertainty Guided Mean Teacher Model for Semi-supervised Medical Image Segmentation',\n",
       "   '10.1007/978-3-030-87196-3_42',\n",
       "   2021),\n",
       "  ('A Persistent Homology-Based Topological Loss Function for Multi-class CNN Segmentation of Cardiac MRI',\n",
       "   '10.1007/978-3-030-68107-4_1',\n",
       "   2020),\n",
       "  ('Have You Forgotten? A Method to Assess if Machine Learning Models Have Forgotten Data',\n",
       "   '10.1007/978-3-030-59710-8_10',\n",
       "   2020),\n",
       "  ('INSIDE: Steering Spatial Attention with Non-imaging Information in CNNs',\n",
       "   '10.1007/978-3-030-59719-1_38',\n",
       "   2020),\n",
       "  ('An Adversarial Network Architecture Using 2D U-Net Models for Segmentation of Left Ventricle from Cine Cardiac MRI',\n",
       "   '10.1007/978-3-030-21949-9_45',\n",
       "   2019),\n",
       "  ('Unsupervised Quality Control of Image Segmentation Based on Bayesian Learning',\n",
       "   '10.1007/978-3-030-32245-8_3',\n",
       "   2019),\n",
       "  ('Cardiac MRI Left Ventricular Segmentation and Function Quantification Using Pre-trained Neural Networks',\n",
       "   '10.1007/978-3-030-78710-3_5',\n",
       "   2021),\n",
       "  ('CNN-Based Cardiac Motion Extraction to Generate Deformable Geometric Left Ventricle Myocardial Models from Cine MRI',\n",
       "   '10.1007/978-3-030-78710-3_25',\n",
       "   2021),\n",
       "  ('Exploring Smoothness and Class-Separation for Semi-supervised Medical Image Segmentation',\n",
       "   '10.1007/978-3-031-16443-9_4',\n",
       "   2022),\n",
       "  ('Left Ventricle Segmentation Using Scale-Independent Multi-Gate UNET in MRI Images',\n",
       "   '10.1007/978-3-030-55789-8_41',\n",
       "   2020),\n",
       "  ('U-Shaped Densely Connected Convolutions for Left Ventricle Segmentation from CMR Images',\n",
       "   '10.1007/978-3-030-89128-2_14',\n",
       "   2021),\n",
       "  ('ShapePU: A New PU Learning Framework Regularized by Global Consistency for Scribble Supervised Cardiac Segmentation',\n",
       "   '10.1007/978-3-031-16452-1_16',\n",
       "   2022),\n",
       "  ('Efficient Bayesian Uncertainty Estimation for nnU-Net',\n",
       "   '10.1007/978-3-031-16452-1_51',\n",
       "   2022),\n",
       "  ('Self-supervised Pretraining for 2D Medical Image Segmentation',\n",
       "   '10.1007/978-3-031-25082-8_31',\n",
       "   2023),\n",
       "  ('Improved AI-Based Segmentation of Apical and Basal Slices from Clinical Cine CMR',\n",
       "   '10.1007/978-3-030-93722-5_10',\n",
       "   2022),\n",
       "  ('LiteTrans: Reconstruct Transformer with Convolution for Medical Image Segmentation',\n",
       "   '10.1007/978-3-030-91415-8_26',\n",
       "   2021),\n",
       "  ('Addressing Class Imbalance in Semi-supervised Image Segmentation: A Study on Cardiac MRI',\n",
       "   '10.1007/978-3-031-16452-1_22',\n",
       "   2022),\n",
       "  ('Evaluating the Performance of StyleGAN2-ADA on Medical Images',\n",
       "   '10.1007/978-3-031-16980-9_14',\n",
       "   2022),\n",
       "  ('Cross Task Temporal Consistency for Semi-supervised Medical Image Segmentation',\n",
       "   '10.1007/978-3-031-21014-3_15',\n",
       "   2022),\n",
       "  ('AMLP-Conv a 3D Axial Long-range Interaction Multilayer Perceptron for CNNs',\n",
       "   '10.1007/978-3-031-21014-3_34',\n",
       "   2022)],\n",
       " 'LA': [('Semi-supervised Left Atrium Segmentation with Mutual Consistency Training',\n",
       "   '10.1007/978-3-030-87196-3_28',\n",
       "   2021),\n",
       "  ('Dual-Task Mutual Learning for Semi-supervised Medical Image Segmentation',\n",
       "   '10.1007/978-3-030-88010-1_46',\n",
       "   2021),\n",
       "  ('Reciprocal Learning for Semi-supervised Segmentation',\n",
       "   '10.1007/978-3-030-87196-3_33',\n",
       "   2021),\n",
       "  ('AtrialGeneral: Domain Generalization for Left Atrial Segmentation of Multi-center LGE MRIs',\n",
       "   '10.1007/978-3-030-87231-1_54',\n",
       "   2021),\n",
       "  ('Exploring Smoothness and Class-Separation for Semi-supervised Medical Image Segmentation',\n",
       "   '10.1007/978-3-031-16443-9_4',\n",
       "   2022),\n",
       "  ('3D Right Ventricle Reconstruction from 2D U-Net Segmentation of Sparse Short-Axis and 4-Chamber Cardiac Cine MRI Views',\n",
       "   '10.1007/978-3-030-93722-5_38',\n",
       "   2022),\n",
       "  ('Decoupling Predictions in Distributed Learning for Multi-center Left Atrial MRI Segmentation',\n",
       "   '10.1007/978-3-031-16431-6_49',\n",
       "   2022),\n",
       "  ('FUSSNet: Fusing Two Sources of Uncertainty for Semi-supervised Medical Image Segmentation',\n",
       "   '10.1007/978-3-031-16452-1_46',\n",
       "   2022),\n",
       "  ('Gabor Filter-Embedded U-Net with Transformer-Based Encoding for Biomedical Image Segmentation',\n",
       "   '10.1007/978-3-031-18814-5_8',\n",
       "   2022),\n",
       "  ('Computerized Analysis of the Human Heart to Guide Targeted Treatment of Atrial Fibrillation',\n",
       "   '10.1007/978-3-031-23443-9_29',\n",
       "   2022)],\n",
       " 'MSCMRSeg': [],\n",
       " 'M&Ms': [],\n",
       " 'PROMISE12': [('Tripled-Uncertainty Guided Mean Teacher Model for Semi-supervised Medical Image Segmentation',\n",
       "   '10.1007/978-3-030-87196-3_42',\n",
       "   2021),\n",
       "  ('Statistical Power in Image Segmentation: Relating Sample Size to Reference Standard Quality',\n",
       "   '10.1007/978-3-319-24574-4_13',\n",
       "   2015),\n",
       "  ('Group Shift Pointwise Convolution for Volumetric Medical Image Segmentation',\n",
       "   '10.1007/978-3-030-87199-4_5',\n",
       "   2021)],\n",
       " 'Medical Decathlon': [],\n",
       " 'I2CVB': [('Segmentation of Prostate in MRI Images Using Depth Separable Convolution Operations',\n",
       "   '10.1007/978-3-030-68449-5_14',\n",
       "   2020)],\n",
       " 'BRATS': [('3D MRI Brain Tumor Segmentation Using Autoencoder Regularization',\n",
       "   '10.1007/978-3-030-11726-9_28',\n",
       "   2018),\n",
       "  ('Brain Tumor Segmentation and Radiomics Survival Prediction: Contribution to the BRATS 2017 Challenge',\n",
       "   '10.1007/978-3-319-75238-9_25',\n",
       "   2017),\n",
       "  ('DeepMedic for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-319-55524-9_14',\n",
       "   2016),\n",
       "  ('Two-Stage Cascaded U-Net: 1st Place Solution to BraTS Challenge 2019 Segmentation Task',\n",
       "   '10.1007/978-3-030-46640-4_22',\n",
       "   2019),\n",
       "  ('TransBTS: Multimodal Brain Tumor Segmentation Using Transformer',\n",
       "   '10.1007/978-3-030-87193-2_11',\n",
       "   2021),\n",
       "  ('nnU-Net for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72087-2_11',\n",
       "   2020),\n",
       "  ('Bag of Tricks for 3D MRI Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-46640-4_20',\n",
       "   2019),\n",
       "  ('3D Dilated Multi-fiber Network for Real-Time Brain Tumor Segmentation in MRI',\n",
       "   '10.1007/978-3-030-32248-9_21',\n",
       "   2019),\n",
       "  ('GLISTRboost: Combining Multimodal MRI Segmentation Registration and Biophysical Tumor Growth Modeling with Gradient Boosting Machines for Glioma Segmentation',\n",
       "   '10.1007/978-3-319-30858-6_13',\n",
       "   2015),\n",
       "  ('Boundary-Aware Fully Convolutional Network for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-319-66185-8_49',\n",
       "   2017),\n",
       "  ('Local Structure Prediction with Convolutional Neural Networks for Multimodal Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-319-42016-5_6',\n",
       "   2015),\n",
       "  ('Deep Convolutional Neural Networks for the Segmentation of Gliomas in Multi-sequence MRI',\n",
       "   '10.1007/978-3-319-30858-6_12',\n",
       "   2015),\n",
       "  ('Automatic Brain Tumor Segmentation Using Convolutional Neural Networks with Test-Time Augmentation',\n",
       "   '10.1007/978-3-030-11726-9_6',\n",
       "   2018),\n",
       "  ('Segmentation of Brain Tumors and Patient Survival Prediction: Methods for the BraTS 2018 Challenge',\n",
       "   '10.1007/978-3-030-11726-9_1',\n",
       "   2018),\n",
       "  ('Brain Tumor Segmentation Using a Fully Convolutional Neural Network with Conditional Random Fields',\n",
       "   '10.1007/978-3-319-55524-9_8',\n",
       "   2016),\n",
       "  ('Brain Tumor Segmentation and Tractographic Feature Extraction from Structural MR Images for Overall Survival Prediction',\n",
       "   '10.1007/978-3-030-11726-9_12',\n",
       "   2018),\n",
       "  ('A Conditional Adversarial Network for Semantic Segmentation of Brain Tumor',\n",
       "   '10.1007/978-3-319-75238-9_21',\n",
       "   2017),\n",
       "  ('Deep Convolutional Neural Networks Using U-Net for Automatic Brain Tumor Segmentation in Multimodal MRI Volumes',\n",
       "   '10.1007/978-3-030-11726-9_4',\n",
       "   2018),\n",
       "  ('3D U-Net Based Brain Tumor Segmentation and Survival Days Prediction',\n",
       "   '10.1007/978-3-030-46640-4_13',\n",
       "   2019),\n",
       "  ('A Partially Reversible U-Net for Memory-Efficient Volumetric Image Segmentation',\n",
       "   '10.1007/978-3-030-32248-9_48',\n",
       "   2019),\n",
       "  ('Brain Lesions Introduction', '10.1007/978-3-319-30858-6_1', 2015),\n",
       "  ('Swin UNETR: Swin Transformers for Semantic Segmentation of Brain Tumors in MRI Images',\n",
       "   '10.1007/978-3-031-08999-2_22',\n",
       "   2022),\n",
       "  ('Brain Tumor Segmentation with Optimized Random Forest',\n",
       "   '10.1007/978-3-319-55524-9_9',\n",
       "   2016),\n",
       "  ('Deep Learning Radiomics Algorithm for Gliomas (DRAG) Model: A Novel Approach Using 3D UNET Based Deep Convolutional Neural Network for Predicting Survival in Gliomas',\n",
       "   '10.1007/978-3-030-11726-9_33',\n",
       "   2018),\n",
       "  ('MRI Brain Tumor Segmentation and Patient Survival Prediction Using Random Forests and Fully Convolutional Networks',\n",
       "   '10.1007/978-3-319-75238-9_18',\n",
       "   2017),\n",
       "  ('3D-ESPNet with Pyramidal Refinement for Volumetric Brain Tumor Image Segmentation',\n",
       "   '10.1007/978-3-030-11726-9_22',\n",
       "   2018),\n",
       "  ('Learning Data Augmentation for Brain Tumor Segmentation with Coarse-to-Fine Generative Adversarial Networks',\n",
       "   '10.1007/978-3-030-11723-8_7',\n",
       "   2018),\n",
       "  ('Deep Recurrent Level Set for Segmenting Brain Tumors',\n",
       "   '10.1007/978-3-030-00931-1_74',\n",
       "   2018),\n",
       "  ('Knowledge Distillation from Multi-modal to Mono-modal Segmentation Networks',\n",
       "   '10.1007/978-3-030-59710-8_75',\n",
       "   2020),\n",
       "  ('Deep Learning Based Multimodal Brain Tumor Diagnosis',\n",
       "   '10.1007/978-3-319-75238-9_13',\n",
       "   2017),\n",
       "  ('Brain Tumor Segmentation and Survival Prediction',\n",
       "   '10.1007/978-3-030-46640-4_32',\n",
       "   2019),\n",
       "  ('Brain Tumor Segmentation Using an Adversarial Network',\n",
       "   '10.1007/978-3-319-75238-9_11',\n",
       "   2017),\n",
       "  ('Vox2Vox: 3D-GAN for Brain Tumour Segmentation',\n",
       "   '10.1007/978-3-030-72084-1_25',\n",
       "   2020),\n",
       "  ('Glioblastoma and Survival Prediction',\n",
       "   '10.1007/978-3-319-75238-9_31',\n",
       "   2017),\n",
       "  ('Focal Dice Loss and Image Dilation for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-00889-5_14',\n",
       "   2018),\n",
       "  ('Brain Tumor Segmentation on MRI with Missing Modalities',\n",
       "   '10.1007/978-3-030-20351-1_32',\n",
       "   2019),\n",
       "  ('Cascaded V-Net Using ROI Masks for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-319-75238-9_33',\n",
       "   2017),\n",
       "  ('Tumor Segmentation and Survival Prediction in Glioma with Deep Learning',\n",
       "   '10.1007/978-3-030-11726-9_8',\n",
       "   2018),\n",
       "  ('Hetero-Modal Variational Encoder-Decoder for Joint Modality Completion and Segmentation',\n",
       "   '10.1007/978-3-030-32245-8_9',\n",
       "   2019),\n",
       "  ('Scribble-Based Hierarchical Weakly Supervised Learning for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-32248-9_20',\n",
       "   2019),\n",
       "  ('Brain Tumor Segmentation with Self-ensembled Deeply-Supervised 3D U-Net Neural Networks: A BraTS 2020 Challenge Solution',\n",
       "   '10.1007/978-3-030-72084-1_30',\n",
       "   2020),\n",
       "  ('Improving Boundary Classification for Brain Tumor Segmentation and Longitudinal Disease Progression',\n",
       "   '10.1007/978-3-319-55524-9_7',\n",
       "   2016),\n",
       "  ('Robust Semantic Segmentation of Brain Tumor Regions from 3D MRIs',\n",
       "   '10.1007/978-3-030-46643-5_8',\n",
       "   2019),\n",
       "  ('Segmentation of Gliomas in Pre-operative and Post-operative Multimodal Magnetic Resonance Imaging Volumes Based on a Hybrid Generative-Discriminative Framework',\n",
       "   '10.1007/978-3-319-55524-9_18',\n",
       "   2016),\n",
       "  ('Context Aware 3D CNNs for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-11726-9_27',\n",
       "   2018),\n",
       "  ('Segmentation of Brain Tumors Using DeepLabv3+',\n",
       "   '10.1007/978-3-030-11726-9_14',\n",
       "   2018),\n",
       "  ('Multi-step Cascaded Networks for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-46640-4_16',\n",
       "   2019),\n",
       "  ('Anatomy-Guided Brain Tumor Segmentation and Classification',\n",
       "   '10.1007/978-3-319-55524-9_16',\n",
       "   2016),\n",
       "  ('3D Deep Neural Network-Based Brain Tumor Segmentation Using Multimodality Magnetic Resonance Sequences',\n",
       "   '10.1007/978-3-319-75238-9_36',\n",
       "   2017),\n",
       "  ('Brain Tumor Segmentation Using Deep Fully Convolutional Neural Networks',\n",
       "   '10.1007/978-3-319-75238-9_30',\n",
       "   2017),\n",
       "  ('Dilated Convolutions for Brain Tumor Segmentation in MRI Scans',\n",
       "   '10.1007/978-3-319-75238-9_22',\n",
       "   2017),\n",
       "  ('Segmentation of Gliomas and Prediction of Patient Overall Survival: A Simple and Fast Procedure',\n",
       "   '10.1007/978-3-030-11726-9_18',\n",
       "   2018),\n",
       "  ('H$$^2$$NF-Net for Brain Tumor Segmentation Using Multimodal MR Imaging: 2nd Place Solution to BraTS Challenge 2020 Segmentation Task',\n",
       "   '10.1007/978-3-030-72087-2_6',\n",
       "   2020),\n",
       "  ('Modality-Pairing Learning for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72084-1_21',\n",
       "   2020),\n",
       "  ('Registration of Pathological Images',\n",
       "   '10.1007/978-3-319-46630-9_10',\n",
       "   2016),\n",
       "  ('Multi-planar Spatial-ConvNet for Segmentation and Survival Prediction in Brain Cancer',\n",
       "   '10.1007/978-3-030-11726-9_9',\n",
       "   2018),\n",
       "  ('Multimodal Brain Tumor Segmentation Using Ensemble of Forest Method',\n",
       "   '10.1007/978-3-319-75238-9_14',\n",
       "   2017),\n",
       "  ('Segmenting Brain Tumors from MRI Using Cascaded Multi-modal U-Nets',\n",
       "   '10.1007/978-3-030-11726-9_2',\n",
       "   2018),\n",
       "  ('Multi-modal PixelNet for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-319-75238-9_26',\n",
       "   2017),\n",
       "  ('Brainlesion: Glioma Multiple Sclerosis Stroke and Traumatic Brain Injuries',\n",
       "   '10.1007/978-3-319-75238-9',\n",
       "   2018),\n",
       "  ('Modality-Aware Mutual Learning for Multi-modal Medical Image Segmentation',\n",
       "   '10.1007/978-3-030-87193-2_56',\n",
       "   2021),\n",
       "  ('Image Features for Brain Lesion Segmentation Using Random Forests',\n",
       "   '10.1007/978-3-319-30858-6_11',\n",
       "   2015),\n",
       "  ('Brain Tumor Segmentation in MRI Scans Using Deeply-Supervised Neural Networks',\n",
       "   '10.1007/978-3-319-75238-9_28',\n",
       "   2017),\n",
       "  ('Multi-resolution 3D CNN for MRI Brain Tumor Segmentation and Survival Prediction',\n",
       "   '10.1007/978-3-030-46640-4_21',\n",
       "   2019),\n",
       "  ('Brain Tumor Synthetic Segmentation in 3D Multimodal MRI Scans',\n",
       "   '10.1007/978-3-030-46640-4_15',\n",
       "   2019),\n",
       "  ('Pooling-Free Fully Convolutional Networks with Dense Skip Connections for Semantic Segmentation with Application to Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-319-75238-9_15',\n",
       "   2017),\n",
       "  ('Deep Hourglass for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-11726-9_37',\n",
       "   2018),\n",
       "  ('Adapting Off-the-Shelf Source Segmenter for Target Medical Image Segmentation',\n",
       "   '10.1007/978-3-030-87196-3_51',\n",
       "   2021),\n",
       "  ('Brain Tumor Segmentation Using Bit-plane and UNET',\n",
       "   '10.1007/978-3-030-11726-9_41',\n",
       "   2018),\n",
       "  ('Automatic Brain Tumor Segmentation with Domain Adaptation',\n",
       "   '10.1007/978-3-030-11726-9_34',\n",
       "   2018),\n",
       "  ('Brain Tumor Classification with Tumor Segmentations and a Dual Path Residual Convolutional Neural Network from MRI and Pathology Images',\n",
       "   '10.1007/978-3-030-46643-5_36',\n",
       "   2019),\n",
       "  ('BiTr-Unet: A CNN-Transformer Combined Network for MRI Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-09002-8_1',\n",
       "   2021),\n",
       "  ('Lifted Auto-Context Forests for Brain Tumour Segmentation',\n",
       "   '10.1007/978-3-319-55524-9_17',\n",
       "   2016),\n",
       "  ('Tumor Segmentation from Multimodal MRI Using Random Forest with Superpixel and Tensor Based Feature Extraction',\n",
       "   '10.1007/978-3-319-75238-9_39',\n",
       "   2017),\n",
       "  ('Extending 2D Deep Learning Architectures to 3D Image Segmentation Problems',\n",
       "   '10.1007/978-3-030-11726-9_7',\n",
       "   2018),\n",
       "  ('Multi-task Learning for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-46640-4_31',\n",
       "   2019),\n",
       "  ('Memory-Efficient Cascade 3D U-Net for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-46640-4_23',\n",
       "   2019),\n",
       "  ('Automatic Brain Tumour Segmentation and Biophysics-Guided Survival Prediction',\n",
       "   '10.1007/978-3-030-46643-5_6',\n",
       "   2019),\n",
       "  ('Automated Brain Tumor Segmentation on Magnetic Resonance Images and Patient’s Overall Survival Prediction Using Support Vector Machines',\n",
       "   '10.1007/978-3-319-75238-9_37',\n",
       "   2017),\n",
       "  ('Binary Glioma Grading: Radiomics versus Pre-trained CNN Features',\n",
       "   '10.1007/978-3-030-00931-1_57',\n",
       "   2018),\n",
       "  ('Multimodal Brain Tumor Segmentation Using Cascaded V-Nets',\n",
       "   '10.1007/978-3-030-11726-9_5',\n",
       "   2018),\n",
       "  ('Brain Tumor Segmentation on Multimodal MR Imaging Using Multi-level Upsampling in Decoder',\n",
       "   '10.1007/978-3-030-11726-9_15',\n",
       "   2018),\n",
       "  ('Intramodality Domain Adaptation Using Self Ensembling and Adversarial Training',\n",
       "   '10.1007/978-3-030-33391-1_4',\n",
       "   2019),\n",
       "  ('Learning High-Resolution and Efficient Non-local Features for Brain Glioma Segmentation in MR Images',\n",
       "   '10.1007/978-3-030-59719-1_47',\n",
       "   2020),\n",
       "  ('Suggestive Annotation of Brain Tumour Images with Gradient-Guided Sampling',\n",
       "   '10.1007/978-3-030-59719-1_16',\n",
       "   2020),\n",
       "  ('Generalized Wasserstein Dice Score Distributionally Robust Deep Learning and Ranger for Brain Tumor Segmentation: BraTS 2020 Challenge',\n",
       "   '10.1007/978-3-030-72087-2_18',\n",
       "   2020),\n",
       "  ('Context Aware 3D UNet for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72084-1_19',\n",
       "   2020),\n",
       "  ('Automated Brain Tumour Segmentation Using Cascaded 3D Densely-Connected U-Net',\n",
       "   '10.1007/978-3-030-72084-1_43',\n",
       "   2020),\n",
       "  ('Hierarchical Multi-class Segmentation of Glioma Images Using Networks with Multi-level Activation Function',\n",
       "   '10.1007/978-3-030-11726-9_11',\n",
       "   2018),\n",
       "  ('Multi-scale Masked 3-D U-Net for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-11726-9_20',\n",
       "   2018),\n",
       "  ('voxel-GAN: Adversarial Framework for Learning Imbalanced Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-11726-9_29',\n",
       "   2018),\n",
       "  ('Multimodal Brain Tumor Segmentation and Survival Prediction Using Hybrid Machine Learning',\n",
       "   '10.1007/978-3-030-46643-5_7',\n",
       "   2019),\n",
       "  ('Brain Tumor Segmentation Based on 3D Residual U-Net',\n",
       "   '10.1007/978-3-030-46643-5_21',\n",
       "   2019),\n",
       "  ('A Two-Stage Cascade Model with Variational Autoencoders and Attention Gates for MRI Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72084-1_39',\n",
       "   2020),\n",
       "  ('Brain Tumor Segmentation and Survival Prediction Using Automatic Hard Mining in 3D CNN Architecture',\n",
       "   '10.1007/978-3-030-72087-2_27',\n",
       "   2020),\n",
       "  ('Automatic Brain Tumor Segmentation with Scale Attention Network',\n",
       "   '10.1007/978-3-030-72084-1_26',\n",
       "   2020),\n",
       "  ('Optimized U-Net for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-09002-8_2',\n",
       "   2021),\n",
       "  ('A Pretrained DenseNet Encoder for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-11726-9_10',\n",
       "   2018),\n",
       "  ('MRI Brain Tumor Segmentation and Uncertainty Estimation Using 3D-UNet Architectures',\n",
       "   '10.1007/978-3-030-72084-1_34',\n",
       "   2020),\n",
       "  ('Multi-stage Deep Layer Aggregation for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72087-2_16',\n",
       "   2020),\n",
       "  ('Brain Tumour Segmentation Using a Triplanar Ensemble of U-Nets on MR Images',\n",
       "   '10.1007/978-3-030-72084-1_31',\n",
       "   2021),\n",
       "  ('Brain Tumor Segmentation from Multimodal MR Images Using Rough Sets',\n",
       "   '10.1007/978-3-319-68124-5_12',\n",
       "   2016),\n",
       "  ('TPCNN: Two-Phase Patch-Based Convolutional Neural Network for Automatic Brain Tumor Segmentation and Survival Prediction',\n",
       "   '10.1007/978-3-319-75238-9_24',\n",
       "   2017),\n",
       "  ('Brain Tumor Segmentation Using a Multi-path CNN Based Method',\n",
       "   '10.1007/978-3-319-75238-9_35',\n",
       "   2017),\n",
       "  ('Glioblastoma Survival Prediction', '10.1007/978-3-030-11726-9_45', 2018),\n",
       "  ('Hybrid Labels for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-46643-5_15',\n",
       "   2019),\n",
       "  ('A Multi-path Decoder Network for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-46643-5_25',\n",
       "   2019),\n",
       "  ('Brain Tumor Segmentation Using Dense Channels 2D U-net and Multiple Feature Extraction Network',\n",
       "   '10.1007/978-3-030-46640-4_26',\n",
       "   2019),\n",
       "  ('DDU-Nets: Distributed Dense Model for 3D MRI Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-46643-5_20',\n",
       "   2019),\n",
       "  ('Knowledge Distillation for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-46643-5_32',\n",
       "   2019),\n",
       "  ('Enhancing MRI Brain Tumor Segmentation with an Additional Classification Network',\n",
       "   '10.1007/978-3-030-72084-1_45',\n",
       "   2020),\n",
       "  ('Impact of Spherical Coordinates Transformation Pre-processing in Deep Convolution Neural Networks for Brain Tumor Segmentation and Survival Prediction',\n",
       "   '10.1007/978-3-030-72084-1_27',\n",
       "   2020),\n",
       "  ('HI-Net: Hyperdense Inception 3D UNet for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72087-2_5',\n",
       "   2020),\n",
       "  ('Extending nn-UNet for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-09002-8_16',\n",
       "   2022),\n",
       "  ('Residual Encoder and Convolutional Decoder Neural Network for Glioma Segmentation',\n",
       "   '10.1007/978-3-319-75238-9_23',\n",
       "   2017),\n",
       "  ('Automatic Segmentation and Overall Survival Prediction in Gliomas Using Fully Convolutional Neural Network and Texture Analysis',\n",
       "   '10.1007/978-3-319-75238-9_19',\n",
       "   2017),\n",
       "  ('Multi-stream Convolutional Autoencoder and 2D Generative Adversarial Network for Glioma Classification',\n",
       "   '10.1007/978-3-030-29888-3_19',\n",
       "   2019),\n",
       "  ('Convolutional 3D to 2D Patch Conversion for Pixel-Wise Glioma Segmentation in MRI Scans',\n",
       "   '10.1007/978-3-030-46640-4_1',\n",
       "   2019),\n",
       "  ('Brain Tumor Segmentation with Cascaded Deep Convolutional Neural Network',\n",
       "   '10.1007/978-3-030-46643-5_9',\n",
       "   2019),\n",
       "  ('Detection and Segmentation of Brain Tumors from MRI Using U-Nets',\n",
       "   '10.1007/978-3-030-46643-5_17',\n",
       "   2019),\n",
       "  ('Brain Tumor Segmentation Network Using Attention-Based Fusion and Spatial Relationship Constraint',\n",
       "   '10.1007/978-3-030-72084-1_20',\n",
       "   2020),\n",
       "  ('ACN: Adversarial Co-training Network for Brain Tumor Segmentation with Missing Modalities',\n",
       "   '10.1007/978-3-030-87234-2_39',\n",
       "   2021),\n",
       "  ('Automatic Brain Tumor Segmentation in Multispectral MRI Volumetric Records',\n",
       "   '10.1007/978-3-319-26561-2_21',\n",
       "   2015),\n",
       "  ('Brain Tumor Segmentation by Variability Characterization of Tumor Boundaries',\n",
       "   '10.1007/978-3-319-55524-9_20',\n",
       "   2016),\n",
       "  ('Brain Tumor Segmentation and Parsing on MRIs Using Multiresolution Neural Networks',\n",
       "   '10.1007/978-3-319-75238-9_29',\n",
       "   2017),\n",
       "  ('3D Large Kernel Anisotropic Network for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-04239-4_40',\n",
       "   2018),\n",
       "  ('A Combined Radio-Histological Approach for Classification of Low Grade Gliomas',\n",
       "   '10.1007/978-3-030-11723-8_42',\n",
       "   2018),\n",
       "  ('An Integrative Analysis of Image Segmentation and Survival of Brain Tumour Patients',\n",
       "   '10.1007/978-3-030-46640-4_35',\n",
       "   2019),\n",
       "  ('Using Separated Inputs for Multimodal Brain Tumor Segmentation with 3D U-Net-like Architectures',\n",
       "   '10.1007/978-3-030-46640-4_18',\n",
       "   2019),\n",
       "  ('Aggregating Multi-scale Prediction Based on 3D U-Net in Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-46640-4_14',\n",
       "   2019),\n",
       "  ('Squeeze-and-Excitation Normalization for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72087-2_32',\n",
       "   2020),\n",
       "  ('Glioma Segmentation Using Ensemble of 2D/3D U-Nets and Survival Prediction Using Multiple Features Fusion',\n",
       "   '10.1007/978-3-030-72087-2_17',\n",
       "   2020),\n",
       "  ('Automatic Segmentation of Brain Tumor Using 3D SE-Inception Networks with Residual Connections',\n",
       "   '10.1007/978-3-030-11726-9_31',\n",
       "   2018),\n",
       "  ('Unsupervised Quality Control of Image Segmentation Based on Bayesian Learning',\n",
       "   '10.1007/978-3-030-32245-8_3',\n",
       "   2019),\n",
       "  ('Ensemble of CNNs for Segmentation of Glioma Sub-regions with Survival Prediction',\n",
       "   '10.1007/978-3-030-46643-5_4',\n",
       "   2019),\n",
       "  ('Brain Tumor Segmentation Based on Attention Mechanism and Multi-model Fusion',\n",
       "   '10.1007/978-3-030-46643-5_5',\n",
       "   2019),\n",
       "  ('Improving Brain Tumor Segmentation with Multi-direction Fusion and Fine Class Prediction',\n",
       "   '10.1007/978-3-030-46640-4_33',\n",
       "   2019),\n",
       "  ('Flow-Based Deformation Guidance for Unpaired Multi-contrast MRI Image-to-Image Translation',\n",
       "   '10.1007/978-3-030-59713-9_70',\n",
       "   2020),\n",
       "  ('DR-Unet104 for Multimodal MRI Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72087-2_36',\n",
       "   2020),\n",
       "  ('Segmentation then Prediction: A Multi-task Solution to Brain Tumor Segmentation and Survival Prediction',\n",
       "   '10.1007/978-3-030-72084-1_44',\n",
       "   2020),\n",
       "  ('Variational-Autoencoder Regularized 3D MultiResUNet for the BraTS 2020 Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72087-2_38',\n",
       "   2020),\n",
       "  ('Modality Completion via Gaussian Process Prior Variational Autoencoders for Multi-modal Glioma Segmentation',\n",
       "   '10.1007/978-3-030-87234-2_42',\n",
       "   2021),\n",
       "  ('ASC-Net: Adversarial-Based Selective Network for Unsupervised Anomaly Segmentation',\n",
       "   '10.1007/978-3-030-87240-3_23',\n",
       "   2021),\n",
       "  ('Automatic Brain Tumor Segmentation in Multispectral MRI Volumes Using a Random Forest Approach',\n",
       "   '10.1007/978-3-319-75786-5_12',\n",
       "   2017),\n",
       "  ('Semi-automatic Brain Tumor Segmentation by Drawing Long Axes on Multi-plane Reformat',\n",
       "   '10.1007/978-3-030-11726-9_39',\n",
       "   2018),\n",
       "  ('Deep Learning Model Integrating Dilated Convolution and Deep Supervision for Brain Tumor Segmentation in Multi-parametric MRI',\n",
       "   '10.1007/978-3-030-32692-0_66',\n",
       "   2019),\n",
       "  ('Multi-modal Segmentation with Missing MR Sequences Using Pre-trained Fusion Networks',\n",
       "   '10.1007/978-3-030-33391-1_19',\n",
       "   2019),\n",
       "  ('A Baseline for Predicting Glioblastoma Patient Survival Time with Classical Statistical Models and Primitive Features Ignoring Image Information',\n",
       "   '10.1007/978-3-030-46640-4_24',\n",
       "   2019),\n",
       "  ('Multi-modal U-Nets with Boundary Loss and Pre-training for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-46643-5_13',\n",
       "   2019),\n",
       "  ('ONCOhabitats Glioma Segmentation Model',\n",
       "   '10.1007/978-3-030-46640-4_28',\n",
       "   2019),\n",
       "  ('Deep Convolutional Neural Networks for Brain Tumor Segmentation: Boosting Performance Using Deep Transfer Learning: Preliminary Results',\n",
       "   '10.1007/978-3-030-46643-5_30',\n",
       "   2019),\n",
       "  ('Multidimensional and Multiresolution Ensemble Networks for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-46643-5_14',\n",
       "   2019),\n",
       "  ('Multimodal Brain Tumor Segmentation and Survival Prediction Using a 3D Self-ensemble ResUNet',\n",
       "   '10.1007/978-3-030-72084-1_33',\n",
       "   2020),\n",
       "  ('MVP U-Net: Multi-View Pointwise U-Net for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72087-2_9',\n",
       "   2020),\n",
       "  ('2D Dense-UNet: A Clinically Valid Approach to Automated Glioma Segmentation',\n",
       "   '10.1007/978-3-030-72087-2_7',\n",
       "   2020),\n",
       "  ('Segmentation Survival Prediction and Uncertainty Estimation of Gliomas from Multimodal 3D MRI Using Selective Kernel Networks',\n",
       "   '10.1007/978-3-030-72087-2_20',\n",
       "   2020),\n",
       "  ('Efficient MRI Brain Tumor Segmentation Using Multi-resolution Encoder-Decoder Networks',\n",
       "   '10.1007/978-3-030-72087-2_3',\n",
       "   2020),\n",
       "  ('MultiATTUNet: Brain Tumor Segmentation and Survival Multitasking',\n",
       "   '10.1007/978-3-030-72084-1_38',\n",
       "   2020),\n",
       "  ('Uncertainty-Driven Refinement of Tumor-Core Segmentation Using 3D-to-2D Networks with Label Uncertainty',\n",
       "   '10.1007/978-3-030-72084-1_36',\n",
       "   2020),\n",
       "  ('An Automatic Overall Survival Time Prediction System for Glioma Brain Tumor Patients Based on Volumetric and Shape Features',\n",
       "   '10.1007/978-3-030-72087-2_31',\n",
       "   2020),\n",
       "  ('Ensemble CNN Networks for GBM Tumors Segmentation Using Multi-parametric MRI',\n",
       "   '10.1007/978-3-031-08999-2_41',\n",
       "   2022),\n",
       "  ('mmFormer: Multimodal Medical Transformer for Incomplete Multimodal Learning of Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-16443-9_11',\n",
       "   2022),\n",
       "  ('Clique Identification and Propagation for Multimodal Brain Tumor Image Segmentation',\n",
       "   '10.1007/978-3-319-47103-7_28',\n",
       "   2016),\n",
       "  ('Towards a Second Brain Images of Tumours for Evaluation (BITE2) Database',\n",
       "   '10.1007/978-3-319-55524-9_2',\n",
       "   2016),\n",
       "  ('Global Planar Convolutions for Improved Context Aggregation in Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-11726-9_35',\n",
       "   2019),\n",
       "  ('Efficient 3D Depthwise and Separable Convolutions with Dilation for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-35288-2_45',\n",
       "   2019),\n",
       "  ('Multimodal Brain Image Segmentation and Analysis with Neuromorphic Attention-Based Learning',\n",
       "   '10.1007/978-3-030-46643-5_2',\n",
       "   2019),\n",
       "  ('Multimodal Segmentation with MGF-Net and the Focal Tversky Loss Function',\n",
       "   '10.1007/978-3-030-46643-5_18',\n",
       "   2019),\n",
       "  ('Fully Automated Brain Tumor Segmentation and Survival Prediction of Gliomas Using Deep Learning and MRI',\n",
       "   '10.1007/978-3-030-46643-5_10',\n",
       "   2019),\n",
       "  ('Automatic Segmentation of Non-tumor Tissues in Glioma MR Brain Images Using Deformable Registration with Partial Convolutional Networks',\n",
       "   '10.1007/978-3-030-72084-1_4',\n",
       "   2020),\n",
       "  ('Brain Tumor Segmentation Using Dual-Path Attention U-Net in 3D MRI Images',\n",
       "   '10.1007/978-3-030-72084-1_17',\n",
       "   2020),\n",
       "  ('Glioma Segmentation Using Encoder-Decoder Network and Survival Prediction Based on Cox Analysis',\n",
       "   '10.1007/978-3-030-72084-1_29',\n",
       "   2020),\n",
       "  ('Glioma Sub-region Segmentation on Multi-parameter MRI with Label Dropout',\n",
       "   '10.1007/978-3-030-72087-2_37',\n",
       "   2020),\n",
       "  ('3D Brain Tumor Segmentation and Survival Prediction Using Ensembles of Convolutional Neural Networks',\n",
       "   '10.1007/978-3-030-72087-2_21',\n",
       "   2020),\n",
       "  ('Multimodal Brain Image Analysis and Survival Prediction Using Neuromorphic Attention-Based Neural Networks',\n",
       "   '10.1007/978-3-030-72084-1_18',\n",
       "   2020),\n",
       "  ('Self-training for Brain Tumour Segmentation with Uncertainty Estimation and Biophysics-Guided Survival Prediction',\n",
       "   '10.1007/978-3-030-72084-1_46',\n",
       "   2020),\n",
       "  ('Brain Tumour Segmentation Using Probabilistic U-Net',\n",
       "   '10.1007/978-3-030-72087-2_22',\n",
       "   2020),\n",
       "  ('Memory Efficient 3D U-Net with Reversible Mobile Inverted Bottlenecks for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72087-2_34',\n",
       "   2021),\n",
       "  ('On the Relationship Between Calibrated Predictors and Unbiased Volume Estimation',\n",
       "   '10.1007/978-3-030-87193-2_64',\n",
       "   2021),\n",
       "  ('A Joint Graph and Image Convolution Network for Automatic Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-08999-2_30',\n",
       "   2021),\n",
       "  ('Prediction of MGMT Methylation Status of Glioblastoma Using Radiomics and Latent Space Shape Features',\n",
       "   '10.1007/978-3-031-09002-8_20',\n",
       "   2021),\n",
       "  ('HNF-Netv2 for Brain Tumor Segmentation Using Multi-modal MR Imaging',\n",
       "   '10.1007/978-3-031-09002-8_10',\n",
       "   2022),\n",
       "  ('AttU-NET: Attention U-Net for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-09002-8_27',\n",
       "   2022),\n",
       "  ('Multimodal Brain Tumor Segmentation Using a 3D ResUNet in BraTS 2021',\n",
       "   '10.1007/978-3-031-08999-2_26',\n",
       "   2022),\n",
       "  ('Reciprocal Adversarial Learning for Brain Tumor Segmentation: A Solution to BraTS Challenge 2021 Segmentation Task',\n",
       "   '10.1007/978-3-031-08999-2_13',\n",
       "   2022),\n",
       "  ('LKAU-Net: 3D Large-Kernel Attention-Based U-Net for Automatic MRI Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-12053-4_24',\n",
       "   2022),\n",
       "  ('TBraTS: Trusted Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-16452-1_48',\n",
       "   2022),\n",
       "  ('Multipath Densely Connected Convolutional Neural Network for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-11723-8_8',\n",
       "   2018),\n",
       "  ('Non-parametric Brain Tissues Segmentation via a Parallel Architecture of CNNs',\n",
       "   '10.1007/978-3-030-21077-9_20',\n",
       "   2019),\n",
       "  ('Combined Learning for Similar Tasks with Domain-Switching Networks',\n",
       "   '10.1007/978-3-030-32254-0_63',\n",
       "   2019),\n",
       "  ('Brain Tumour Segmentation from Multispectral MR Image Data Using Ensemble Learning Methods',\n",
       "   '10.1007/978-3-030-33904-3_30',\n",
       "   2019),\n",
       "  ('A Study on Histogram Normalization for Brain Tumour Segmentation from Multispectral MR Image Data',\n",
       "   '10.1007/978-3-030-33904-3_35',\n",
       "   2019),\n",
       "  ('Improving Brain Tumor Segmentation with Dilated Pseudo-3D Convolution and Multi-direction Fusion',\n",
       "   '10.1007/978-3-030-37731-1_59',\n",
       "   2020),\n",
       "  ('cuRadiomics: A GPU-Based Radiomics Feature Extraction Toolkit',\n",
       "   '10.1007/978-3-030-40124-5_5',\n",
       "   2019),\n",
       "  ('Improving Brain Tumor Segmentation in Multi-sequence MR Images Using Cross-Sequence MR Image Generation',\n",
       "   '10.1007/978-3-030-46643-5_3',\n",
       "   2019),\n",
       "  ('The Tumor Mix-Up in 3D Unet for Glioma Segmentation',\n",
       "   '10.1007/978-3-030-46643-5_26',\n",
       "   2019),\n",
       "  ('End-to-End Nested Multi-Attention Network for 3D Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-59585-2_6',\n",
       "   2020),\n",
       "  ('AutoSyncoder: An Adversarial AutoEncoder Framework for Multimodal MRI Synthesis',\n",
       "   '10.1007/978-3-030-61598-7_10',\n",
       "   2020),\n",
       "  ('Brain Tumor Segmentation from Multi-spectral MR Image Data Using Random Forest Classifier',\n",
       "   '10.1007/978-3-030-63830-6_15',\n",
       "   2020),\n",
       "  ('An Empirical Study of Deep Neural Networks for Glioma Detection from MRI Sequences',\n",
       "   '10.1007/978-3-030-63830-6_10',\n",
       "   2020),\n",
       "  ('TwoPath U-Net for Automatic Brain Tumor Segmentation from Multimodal MRI Data',\n",
       "   '10.1007/978-3-030-72087-2_26',\n",
       "   2020),\n",
       "  ('A Deep Supervision CNN Network for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72087-2_14',\n",
       "   2020),\n",
       "  ('Segmenting Brain Tumors from MRI Using Cascaded 3D U-Nets',\n",
       "   '10.1007/978-3-030-72087-2_23',\n",
       "   2020),\n",
       "  ('E1D3 U-Net for Brain Tumor Segmentation: Submission to the RSNA-ASNR-MICCAI BraTS 2021 challenge',\n",
       "   '10.1007/978-3-031-09002-8_25',\n",
       "   2021),\n",
       "  ('Generator Versus Segmentor: Pseudo-healthy Synthesis',\n",
       "   '10.1007/978-3-030-87231-1_15',\n",
       "   2021),\n",
       "  ('Redundancy Reduction in Semantic Segmentation of 3D Brain Tumor MRIs',\n",
       "   '10.1007/978-3-031-09002-8_15',\n",
       "   2021),\n",
       "  ('Evidence Fusion with Contextual Discounting for Multi-modality Medical Image Segmentation',\n",
       "   '10.1007/978-3-031-16443-9_39',\n",
       "   2022),\n",
       "  ('An Ensemble Approach to Automatic Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-09002-8_13',\n",
       "   2022),\n",
       "  ('Multi-plane UNet++ Ensemble for Glioblastoma Segmentation',\n",
       "   '10.1007/978-3-031-08999-2_23',\n",
       "   2022),\n",
       "  ('Brain Tumor Segmentation Using Neural Network Topology Search',\n",
       "   '10.1007/978-3-031-08999-2_31',\n",
       "   2022),\n",
       "  ('Diffusion Models for Medical Anomaly Detection',\n",
       "   '10.1007/978-3-031-16452-1_4',\n",
       "   2022),\n",
       "  ('msFormer: Adaptive Multi-Modality 3D Transformer for Medical Image Segmentation',\n",
       "   '10.1007/978-3-031-18910-4_26',\n",
       "   2022),\n",
       "  ('Interactive Semi-automated Method Using Non-negative Matrix Factorization and Level Set Segmentation for the BRATS Challenge',\n",
       "   '10.1007/978-3-319-55524-9_19',\n",
       "   2016),\n",
       "  ('Semi-supervised Variational Autoencoder for Survival Prediction',\n",
       "   '10.1007/978-3-030-46643-5_12',\n",
       "   2019),\n",
       "  ('Children’s Neuroblastoma Segmentation Using Morphological Features',\n",
       "   '10.1007/978-3-030-32692-0_10',\n",
       "   2019),\n",
       "  ('Multi-branch Learning Framework with Different Receptive Fields Ensemble for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-46643-5_27',\n",
       "   2019),\n",
       "  ('Two Stages CNN-Based Segmentation of Gliomas Uncertainty Quantification and Prediction of Overall Patient Survival',\n",
       "   '10.1007/978-3-030-46643-5_16',\n",
       "   2019),\n",
       "  ('Multimodal Brain Tumor Segmentation with Normal Appearance Autoencoder',\n",
       "   '10.1007/978-3-030-46643-5_31',\n",
       "   2019),\n",
       "  ('Brain Tumour Segmentation on MRI Images by Voxel Classification Using Neural Networks and Patient Survival Prediction',\n",
       "   '10.1007/978-3-030-46640-4_27',\n",
       "   2019),\n",
       "  ('Brain Tumor Segmentation Using 3D Convolutional Neural Network',\n",
       "   '10.1007/978-3-030-46643-5_19',\n",
       "   2019),\n",
       "  ('Error Attention Interactive Segmentation of Medical Image Through Matting and Fusion',\n",
       "   '10.1007/978-3-030-59861-7_2',\n",
       "   2020),\n",
       "  ('Overall Survival Prediction in Gliomas Using Region-Specific Radiomic Features',\n",
       "   '10.1007/978-3-030-66843-3_25',\n",
       "   2020),\n",
       "  ('Efficient Brain Tumour Segmentation Using Co-registered Data and Ensembles of Specialised Learners',\n",
       "   '10.1007/978-3-030-72087-2_2',\n",
       "   2020),\n",
       "  ('Learning Dynamic Convolutions for Multi-modal 3D MRI Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72087-2_39',\n",
       "   2020),\n",
       "  ('A Deep Supervised U-Attention Net for Pixel-Wise Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72087-2_24',\n",
       "   2020),\n",
       "  ('A Deep Random Forest Approach for Multimodal Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72087-2_12',\n",
       "   2020),\n",
       "  ('Glioma Segmentation with 3D U-Net Backed with Energy-Based Post-Processing',\n",
       "   '10.1007/978-3-030-72087-2_10',\n",
       "   2020),\n",
       "  ('Low-Rank Convolutional Networks for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72084-1_42',\n",
       "   2020),\n",
       "  ('Efficient Embedding Network for 3D Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72084-1_23',\n",
       "   2020),\n",
       "  ('Some New Tricks for Deep Glioma Segmentation',\n",
       "   '10.1007/978-3-030-72087-2_28',\n",
       "   2020),\n",
       "  ('MIST GAN: Modality Imputation Using Style Transfer for MRI',\n",
       "   '10.1007/978-3-030-87589-3_34',\n",
       "   2021),\n",
       "  ('Group Shift Pointwise Convolution for Volumetric Medical Image Segmentation',\n",
       "   '10.1007/978-3-030-87199-4_5',\n",
       "   2021),\n",
       "  ('Cross-modality Attention Method for Medical Image Enhancement',\n",
       "   '10.1007/978-3-030-88010-1_34',\n",
       "   2021),\n",
       "  ('Multidimensional and Multiresolution Ensemble Networks for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-030-72084-1_40',\n",
       "   2021),\n",
       "  ('ACT: Semi-supervised Domain-Adaptive Medical Image Segmentation with Asymmetric Co-training',\n",
       "   '10.1007/978-3-031-16443-9_7',\n",
       "   2022),\n",
       "  ('Weighted Metamorphosis for Registration of Images with Different Topologies',\n",
       "   '10.1007/978-3-031-11203-4_2',\n",
       "   2022),\n",
       "  ('Quality-Aware Model Ensemble for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-09002-8_14',\n",
       "   2022),\n",
       "  ('Generalized Wasserstein Dice Loss Test-Time Augmentation and Transformers for the BraTS 2021 Challenge',\n",
       "   '10.1007/978-3-031-09002-8_17',\n",
       "   2022),\n",
       "  ('Brain Tumor Segmentation from Multiparametric MRI Using a Multi-encoder U-Net Architecture',\n",
       "   '10.1007/978-3-031-09002-8_26',\n",
       "   2022),\n",
       "  ('Automatic Brain Tumor Segmentation with a Bridge-Unet Deeply Supervised Enhanced with Downsampling Pooling Combination Atrous Spatial Pyramid Pooling Squeeze-and-Excitation and EvoNorm',\n",
       "   '10.1007/978-3-031-09002-8_23',\n",
       "   2022),\n",
       "  ('Combining CNNs with Transformer for Multimodal 3D MRI Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-09002-8_21',\n",
       "   2022),\n",
       "  ('Dice Focal Loss with ResNet-like Encoder-Decoder Architecture in 3D Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-09002-8_9',\n",
       "   2022),\n",
       "  ('Comparison of MR Preprocessing Strategies and Sequences for Radiomics-Based MGMT Prediction',\n",
       "   '10.1007/978-3-031-09002-8_33',\n",
       "   2022),\n",
       "  ('MS UNet: Multi-scale 3D UNet for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-09002-8_3',\n",
       "   2022),\n",
       "  ('Feature Learning by Attention and Ensemble with 3D U-Net to Glioma Tumor Segmentation',\n",
       "   '10.1007/978-3-031-09002-8_6',\n",
       "   2022),\n",
       "  ('Disparity Autoencoders for Multi-class Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-09002-8_11',\n",
       "   2022),\n",
       "  ('Neural Network Based Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-09002-8_29',\n",
       "   2022),\n",
       "  ('MRI Brain Tumor Segmentation Using Deep Encoder-Decoder Convolutional Neural Networks',\n",
       "   '10.1007/978-3-031-09002-8_7',\n",
       "   2022),\n",
       "  ('BRATS2021: Exploring Each Sequence in Multi-modal Input for Baseline U-net Performance',\n",
       "   '10.1007/978-3-031-08999-2_15',\n",
       "   2022),\n",
       "  ('A Two-Phase Optimal Mass Transportation Technique for 3D Brain Tumor Detection and Segmentation',\n",
       "   '10.1007/978-3-031-08999-2_34',\n",
       "   2022),\n",
       "  ('Multi Modal Fusion for Radiogenomics Classification of Brain Tumor',\n",
       "   '10.1007/978-3-031-08999-2_29',\n",
       "   2022),\n",
       "  ('3D MRI Brain Tumour Segmentation with Autoencoder Regularization and Hausdorff Distance Loss Function',\n",
       "   '10.1007/978-3-031-08999-2_27',\n",
       "   2022),\n",
       "  ('Ensemble Outperforms Single Models in Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-08999-2_39',\n",
       "   2022),\n",
       "  ('Simple and Fast Convolutional Neural Network Applied to Median Cross Sections for Predicting the Presence of MGMT Promoter Methylation in FLAIR MRI Scans',\n",
       "   '10.1007/978-3-031-08999-2_18',\n",
       "   2022),\n",
       "  ('NnUNet with Region-based Training and Loss Ensembles for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-08999-2_36',\n",
       "   2022),\n",
       "  ('Automatic Brain Tumor Segmentation Using Multi-scale Features and Attention Mechanism',\n",
       "   '10.1007/978-3-031-08999-2_17',\n",
       "   2022),\n",
       "  ('Hierarchical and Global Modality Interaction for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-08999-2_38',\n",
       "   2022),\n",
       "  ('Brain Tumor Segmentation Using UNet-Context Encoding Network',\n",
       "   '10.1007/978-3-031-08999-2_40',\n",
       "   2022),\n",
       "  ('Segmenting Brain Tumors in Multi-modal MRI Scans Using a 3D SegNet Architecture',\n",
       "   '10.1007/978-3-031-08999-2_32',\n",
       "   2022),\n",
       "  ('HarDNet-BTS: A Harmonic Shortcut Network for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-08999-2_21',\n",
       "   2022),\n",
       "  ('A 3D Medical Image Segmentation Framework Fusing Convolution and Transformer Features',\n",
       "   '10.1007/978-3-031-13870-6_63',\n",
       "   2022),\n",
       "  ('NestedFormer: Nested Modality-Aware Transformer for Brain Tumor Segmentation',\n",
       "   '10.1007/978-3-031-16443-9_14',\n",
       "   2022),\n",
       "  ('Uncertainty Categories in Medical Image Segmentation: A Study of Source-Related Diversity',\n",
       "   '10.1007/978-3-031-16749-2_3',\n",
       "   2022),\n",
       "  ('Med-DANet: Dynamic Architecture Network for Efficient Medical Volumetric Segmentation',\n",
       "   '10.1007/978-3-031-19803-8_30',\n",
       "   2022),\n",
       "  ('Human Knowledge-Guided and Task-Augmented Deep Learning for Glioma Grading',\n",
       "   '10.1007/978-3-031-18910-4_45',\n",
       "   2022),\n",
       "  ('Unsupervised Anomaly Segmentation for Brain Lesions Using Dual Semantic-Manifold Reconstruction',\n",
       "   '10.1007/978-3-031-30111-7_12',\n",
       "   2023)],\n",
       " 'Synapse': []}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for d in paper_using_abstract:\n",
    "    print(f\"Number of citations for {d}: {len(paper_using_abstract[d])}\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Check tables in fulltext for dataset's name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Check for dataset's organ in figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check fulltext of paper for either figures or tables"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import libs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#To download fulltext\n",
    "import requests\n",
    "import pandas as pd\n",
    "\n",
    "#To filter invalid pdf\n",
    "from pypdf import PdfReader\n",
    "from pypdf.errors import PdfReadError\n",
    "\n",
    "#To handle files\n",
    "import glob\n",
    "import os\n",
    "\n",
    "#To extract images from pdf\n",
    "import fitz\n",
    "\n",
    "#To extract tables from pdf\n",
    "import camelot\n",
    "import ghostscript"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download papers full text"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter paper to get for the test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Only get paper from 2023 referencing ACDC paper\n",
    "df = pd.read_csv(\"../../results/extracted_csv/paper_openalex.csv\")\n",
    "df_2023 = df[df[\"publication_year\"] == 2023]\n",
    "df_acdc = df_2023[df_2023[\"dataset_used\"] == \"ACDC\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10.1016/b978-0-32-385773-4.00023-x\n",
      "10.1016/b978-0-32-385773-4.00025-3\n",
      "10.1016/b978-0-32-385773-4.00009-5\n",
      "https://doi.org/10.1016/b978-0-32-385773-4.00009-5\n",
      "10.1016/j.compbiomed.2022.106439\n",
      "10.1371/journal.pdig.0000159\n",
      "https://journals.plos.org/digitalhealth/article/file?id=10.1371/journal.pdig.0000159&type=printable\n",
      "10.1109/access.2023.3234241\n",
      "10.1016/j.compmedimag.2022.102174\n",
      "10.1088/1361-6560/acb19a\n",
      "https://iopscience.iop.org/article/10.1088/1361-6560/acb19a/pdf\n",
      "10.1007/s00740-022-00474-9\n",
      "10.1016/j.patcog.2023.109318\n",
      "https://doi.org/10.1016/j.patcog.2023.109318\n",
      "10.1109/access.2023.3238058\n",
      "10.1038/s41598-023-28348-y\n",
      "https://www.nature.com/articles/s41598-023-28348-y.pdf\n",
      "10.3389/fphys.2023.1027076\n",
      "https://www.frontiersin.org/articles/10.3389/fphys.2023.1027076/pdf\n",
      "10.1016/b978-0-12-821983-6.00008-4\n",
      "10.3390/bioengineering10020166\n",
      "https://www.mdpi.com/2306-5354/10/2/166/pdf?version=1674889337\n",
      "10.1016/j.media.2023.102762\n",
      "http://arxiv.org/pdf/2206.01136\n",
      "10.1016/j.bspc.2023.104631\n",
      "10.1109/wacv56688.2023.00365\n",
      "https://eprints.gla.ac.uk/282738/2/282738.pdf\n"
     ]
    }
   ],
   "source": [
    "url_base = \"https://api.openalex.org/works/https://doi.org/\"\n",
    "paper_id = 1\n",
    "for doi in df_acdc[\"DOI\"]:\n",
    "    url = url_base + doi \n",
    "    r_paper = requests.get(url)\n",
    "    if r_paper.status_code == 200:\n",
    "        r_paper_json = r_paper.json()\n",
    "        fulltext_url = r_paper_json[\"open_access\"][\"oa_url\"]\n",
    "        if fulltext_url:\n",
    "            r_fulltext = requests.get(fulltext_url,allow_redirects=True)\n",
    "            if r_fulltext.status_code == 200:\n",
    "                print(fulltext_url)\n",
    "                open(f\"../../results/papers_fulltext/{paper_id}.pdf\",\"wb\").write(r_fulltext.content)\n",
    "                paper_id += 1\n",
    "        #Stop after 10th download, only to test and maybe not get block by some site\n",
    "        if paper_id == 10:\n",
    "            break\n",
    "    else:\n",
    "        print(f\"Error {r_paper.status_code} for {doi}\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove invalid pdf obtain from previous step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get every downloaded pdf path\n",
    "pdf_path = glob.glob(\"../../results/papers_fulltext/*.pdf\")\n",
    "\n",
    "#Create folder where valid pdf will be moved\n",
    "if not os.path.exists(\"../../results/papers_fulltext/valid_pdf\"):\n",
    "    os.makedirs(\"../../results/papers_fulltext/valid_pdf\")\n",
    "\n",
    "#Create folder where invalid pdf will be moved\n",
    "if not os.path.exists(\"../../results/papers_fulltext/removed_pdf\"):\n",
    "    os.makedirs(\"../../results/papers_fulltext/removed_pdf\")\n",
    "\n",
    "#For each downloaded pdf\n",
    "for file in pdf_path:\n",
    "    try:\n",
    "        #Try to read the pdf (Raise an error if the file is an invalid pdf)\n",
    "        PdfReader(file,strict=True)\n",
    "        #If valid, move the file to valid folder\n",
    "        os.rename(file, file.replace(\"papers_fulltext/\",\"papers_fulltext/valid_pdf/\"))\n",
    "    except PdfReadError:\n",
    "        #If a PdfReadError is raised, the pdf is invalid and therefor moved to removed_pdf folder\n",
    "        os.rename(file, file.replace(\"papers_fulltext/\",\"papers_fulltext/removed_pdf/\"))\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract images from valid pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get every valid pdf path\n",
    "valid_pdf_path = glob.glob(\"../../results/papers_fulltext/valid_pdf/*.pdf\")\n",
    "\n",
    "#path of folder where images will be stored\n",
    "images_path = \"../../results/papers_fulltext/images\"\n",
    "#Create this folder if it does not exist\n",
    "if not os.path.exists(images_path):\n",
    "    os.makedirs(images_path)\n",
    "\n",
    "#For each pdf file\n",
    "for file in valid_pdf_path:\n",
    "    #Open the file\n",
    "    pdf_file = fitz.open(file)\n",
    "\n",
    "    #Get the number of pages in PDF file\n",
    "    page_nums = len(pdf_file)\n",
    "\n",
    "    #Create empty list to store images information\n",
    "    images_list = []\n",
    "\n",
    "    #Extract all images information from each page\n",
    "    for page_num in range(page_nums):\n",
    "        page_content = pdf_file[page_num]\n",
    "        images_list.extend(page_content.get_images())\n",
    "\n",
    "    #If there is at least one image in the pdf\n",
    "    if len(images_list)!=0:\n",
    "        #Create a subfolder for the article, this way we easily know from which paper the images is coming from\n",
    "        if not os.path.exists(os.path.join(images_path, os.path.basename(file.replace(\".pdf\",\"\")))):\n",
    "            os.makedirs(os.path.join(images_path, os.path.basename(file.replace(\".pdf\",\"\"))))\n",
    "\n",
    "        #Save all the extracted images\n",
    "        for i, img in enumerate(images_list, start=1):\n",
    "            #Extract the image object number\n",
    "            xref = img[0]\n",
    "            #Extract image\n",
    "            base_image = pdf_file.extract_image(xref)\n",
    "            #Store image bytes\n",
    "            image_bytes = base_image['image']\n",
    "            #Store image extension\n",
    "            image_ext = base_image['ext']\n",
    "            #Generate image file name\n",
    "            image_name = str(i) + '.' + image_ext\n",
    "            #Save image\n",
    "            with open(os.path.join(images_path, os.path.basename(file.replace(\".pdf\",\"\")),image_name) , 'wb') as image_file:\n",
    "                image_file.write(image_bytes)\n",
    "                image_file.close()\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract tables from valid PDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "valid_pdf_path = glob.glob(\"../../results/papers_fulltext/valid_pdf/*.pdf\")\n",
    "for pdf in valid_pdf_path:\n",
    "    tables = camelot.read_pdf(pdf,\"all\",flavor=\"stream\",suppress_stdout=True)\n",
    "    for t in tables:\n",
    "        t_str = t.df.to_string()\n",
    "        if \"ACDC\" in t_str:\n",
    "            print(\"ACDC IN\",pdf)\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACDC IN ../../results/papers_fulltext/valid_pdf/5.pdf\n",
      "ACDC IN ../../results/papers_fulltext/valid_pdf/9.pdf\n",
      "ACDC IN ../../results/papers_fulltext/valid_pdf/6.pdf\n",
      "ACDC IN ../../results/papers_fulltext/valid_pdf/7.pdf\n",
      "ACDC IN ../../results/papers_fulltext/valid_pdf/8.pdf\n"
     ]
    }
   ],
   "source": [
    "for pdf in valid_pdf_path:\n",
    "    tables = camelot.read_pdf(pdf,\"all\",flavor=\"stream\",suppress_stdout=True)\n",
    "    for t in tables:\n",
    "        t_str = t.df.to_string()\n",
    "        if \"ACDC\" in t_str:\n",
    "            print(\"ACDC IN\",pdf)\n",
    "            break"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
